File: src/13-challenge/solution.py
import re
from typing import Literal

type Move = Literal["D", "L", "R", "U"]


def is_robot_back(moves: str) -> Literal[True] | list[int]:
    current_pos: list[int] = [0, 0]

    opposite_moves: dict[Move, Move] = {
        "D": "U",
        "L": "R",
        "R": "L",
        "U": "D",
    }

    move_actions: dict[Move, list[int]] = {
        "D": [0, -1],
        "L": [-1, 0],
        "R": [1, 0],
        "U": [0, 1],
    }

    moves = re.sub(
        pattern=r"[*][DLRU]",
        repl=lambda value: value.group(0)[1] * 2,
        string=moves,
    )

    moves = re.sub(
        pattern=r"[!][DLRU]",
        repl=lambda value: opposite_moves[value.group(0)[1]],  # type: ignore
        string=moves,
    )

    aux: str = ""
    for char in list(moves):
        if aux.find("?") != -1:
            aux = aux.replace("?", "")
            aux += char if aux.find(char) == -1 else ""
        else:
            aux += char

    moves = aux

    for move in moves:
        offset: list[int] = move_actions[move]  # type: ignore
        current_pos[0] += offset[0]
        current_pos[1] += offset[1]

    return current_pos if current_pos[0] + current_pos[1] else True

File: src/13-challenge/solution.js
/**
 * @param {string} moves
 * @returns {true|[number, number]} Return true if robot returns or position
 */
export default function isRobotBack(moves) {
	/** @type {[number, number]} */
	const currentPos = [0, 0]

	/** @type {Object.<'D' | 'L' | 'R' | 'U', 'D' | 'L' | 'R' | 'U'>} */
	const oppositeMoves = {
		D: 'U',
		L: 'R',
		R: 'L',
		U: 'D',
	}

	/** @type {Object.<'D' | 'L' | 'R' | 'U', [number, number]>} */
	const moveActions = {
		D: [0, -1],
		L: [-1, 0],
		R: [1, 0],
		U: [0, 1],
	}

	moves = moves.replaceAll(/[*][DLRU]/g, (value) => value.substring(1).repeat(2))
	moves = moves.replaceAll(/[!][DLRU]/g, (value) => oppositeMoves[value.substring(1)])

	moves = moves.split('').reduce((prev, current) => {
		if (prev.includes('?')) {
			prev = prev.replaceAll('?', '')
			return !prev.includes(current) ? prev + current : prev
		}

		return prev + current
	})

	for (const move of moves) {
		/** @type {[number, number]}>} */
		const offset = moveActions[move]
		currentPos[0] += offset[0]
		currentPos[1] += offset[1]
	}

	return currentPos[0] + currentPos[1] ? currentPos : true
}
